### PR TITLE
CORE: config file support

### PR DESCRIPTION
**Pull request by vspetrov**
## What
Adds support for configuration files with UCC parameters

## Why ?
Allows distribution of UCC with pre-built configuration

## How ?
configuration file can contain any UCC variables in the format var = value (";" symbol in the begging of the file comments out the setting). Configuration file is searched by the library once during ucc_constructor call: firstly the env variable UCC_CONFIG_FILE is checked (it has highest precedence); if not found the default file "ucc.conf" is checked in the $HOME; if not found the default file "ucc.conf" is checked in the <ucc_install_dir>/share/ucc.conf.

Example config file:
UCC_TL_UCP_LOG_LEVEL=info
UCC_TLS=ucp
MPI_UCC_TL_UCP_KN_RADIX=3 ;; MPI_ prefix means that the var only affects UCC lib objects with "MPI" env prefix

NOTES:
why "filename" arguments from ucc_lib_confi_read and ucc_context_config_read are not used ?
main problem is user experience: it is not clear for user which variable belongs to "lib" and which to "context", ie user will need to maintain 2 different files and understand which variable to put into each file.
Having 1 global file makes it much more usable imho. 

another reason: ec/mc configrations (e.g. mpool settings). these components are singletons (ie initialized only once, and they are global objects for all ucc_lib objects in the memory space of a process), and their configurations must be global, so can't be read from lib/context specific "files". 

If required we can extend the support for those "filename" args in lib/context_config_read but they will only be applied for some set of variables. I would postpone that until explicitly requested by anyone.

### CODE REVIEW

**Review by Sergei-Lebedev — COMMENTED** [review]
i would add some tests to gtest for the parses to check different corner cases

File: src/core/ucc_lib.c [hunk 1]
```diff
@@ -299,9 +300,6 @@ ucc_status_t ucc_init_version(unsigned api_major_version,
 
     *lib_p = NULL;
 
-    if (UCC_OK != (status = ucc_constructor())) {
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/core/ucc_lib.c:302]: agree: keeping ucc_constructor in both ucc_init_version and lib_config_read. It is singleton, so will be called once from either location

File: src/utils/ucc_parser.h [hunk 2]
```diff
@@ -153,4 +142,9 @@ ucc_status_t ucc_config_allow_list_process(const ucc_config_allow_list_t * list,
                                            const ucc_config_names_array_t *all,
                                            ucc_config_names_list_t *       out);
 
+typedef struct ucc_file_config ucc_file_config_t;
+ucc_status_t                   ucc_parse_file_config(const char *        filename,
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/utils/ucc_parser.h:146]: yes, looks weird. this what out "style" produced. apparently, it aligned on ucc_file_config_t typedef. I moved it to another place and re-formatted.

**Comment by vspetrov** [comment]
@manjugv @bureddy plz have a look

**Review by Sergei-Lebedev — APPROVED** [review]

**Review by manjugv — APPROVED** [review]
